source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2534 @@
+
+- ``documents``: Required array of objects. List of documents corresponding to
+  the contents of the collection. This list may be empty.
+
+
+test
+~~~~
+
+Test case consisting of a sequence of operations to be executed.
+
+The structure of this object is as follows:
+
+- ``description``: Required string. The name of the test.
+
+  This SHOULD describe the purpose of this test (e.g. "insertOne is retried").
+
+.. _test_runOnRequirements:
+
+- ``runOnRequirements``: Optional array of on or more `runOnRequirement`_
+  objects. List of server version and/or topology requirements for which this
+  test can be run. If specified, these requirements are evaluated independently
+  and in addition to any top-level `runOnRequirements`_. If no requirements in
+  this array are met, the test runner MUST skip this test.
+
+  These requirements SHOULD be more restrictive than those specified in the
+  top-level `runOnRequirements`_ (if any). They SHOULD NOT be more permissive.
+  This is advised because both sets of requirements MUST be satisified in order
+  for a test to be executed and more permissive requirements at the test-level
+  could be taken out of context on their own.
+
+.. _test_skipReason:
+
+- ``skipReason``: Optional string. If set, the test will be skipped. The string
+  SHOULD explain the reason for skipping the test (e.g. JIRA ticket).
+
+.. _test_operations:
+
+- ``operations``: Required array of one or more `operation`_ objects. List of
+  operations to be executed for the test case.
+
+.. _test_expectEvents:
+
+- ``expectEvents``: Optional array of one or more `expectedEventsForClient`_
+  objects. For one or more clients, a list of events that are expected to be
+  observed in a particular order.
+
+  If a driver only supports configuring event listeners globally (for all
+  clients), the test runner SHOULD associate each observed event with a client in
+  in order to perform these assertions.
+
+  Test files SHOULD NOT expect events from multiple specs (e.g. command
+  monitoring *and* SDAM events) for a single client. See
+  `Mixing event types in observeEvents and expectEvents`_ for more
+  information.
+
+.. _test_outcome:
+
+- ``outcome``: Optional array of documents. Each document will specify expected
+  contents of a collection after all operations have been executed. The list of
+  documents therein SHOULD be sorted ascendingly by the ``_id`` field to allow
+  for deterministic comparisons.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `collectionData`_.
+
+
+operation
+~~~~~~~~~
+
+An operation to be executed as part of the test.
+
+The structure of this object is as follows:
+
+.. _operation_name:
+
+- ``name``: Required string. Name of the operation (e.g. method) to perform on
+  the object.
+
+.. _operation_object:
+
+- ``object``: Required string. Name of the object on which to perform the
```

Comment:
Is it safe to say that this will always be a `collection` object? It doesn't specify this here.

If not I am slightly concerned by not knowing the `type` of the parent from within the operation.

One reason is for type checking. If there was something in this object that told me what this entity type was, I could use that.

The second reason is argument mapping, if the object type for one operation was `Bar` and the object for another operation was `Baz` but they both had the same method `insertOne` there would be no way for me to programmatically differentiate the mapping.

Reply:
This can be _any_ entity. The "Entity Test Operations" section enumerates various types of entities that may be supported, and links to relevant specs where those methods might be defined.

In some cases, we have custom operations. For example:

 * `createChangeStream` on `client`, `database`, and `collection`
 * `iterateUntilDocumentOrError` on `changeStream`

"Executing an Operation" discusses how to interpret this:

> If `operation.object` is not "testRunner", this is an entity operation. If `operation.object` is defined in the current test's Entity Map, the test runner MUST fetch that entity and note its type; otherwise, the test runner MUST raise an error for an undefined entity. If `operation.name` does not correspond to an operation for the entity type (per Entity Test Operations),
the test runner MUST raise an error for an undefined operation. Test runners MAY skip tests that include operations that are intentionally unimplemented (e.g. ``listCollectionNames``).

This should not be entirely unfamiliar, as existing test runners already different entity types. The only difference is that there is generally only a single instance of each driver class (e.g. "collection", "database") apart from sessions, which could be "session0" or "session1". Either way, the object needs to get resolved to a type in order to interpret the method name and arguments. In this case, the implementation just asks that you resolve the type via the entity map instead of hard-coding one of a few known strings. I would imagine most drivers already implement this via some sort of `switch` or conditional sequence.

To use PHPLIB as an example (https://github.com/mongodb/mongo-php-library/blob/master/tests/SpecTests/Operation.php#L292), I expect that `switch` will be modified to instead look up the object in the entity map and check its class instead of comparing against a set of known constants.

Reply:
Hey thanks for clarifying this. I think your example of checking the class type / instance of the `object` makes sense, and eliminates the need to specify a parent type.